### PR TITLE
Quick fix for broken used name

### PR DESCRIPTION
**Pull request by jilen**
Fix issue mentioned here https://github.com/sbt/sbt/issues/3666

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?zinc=pull/449/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/252/console">is in progress...</a>

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala [hunk 1]
```diff
@@ -605,7 +605,7 @@ private class MRelationsNameHashing(
       internalDependencies = internalDependencies,
       externalDependencies = externalDependencies,
       classes,
-      names = names + (className, name),
+      names = names + (className, name.replaceAllLiterally("\n", "")),
```
Comment (jvican) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala]: Can you hoist up this expression into a function that is invoked in a hoisted up value?
  
Also, this patch makes both a name with a new line and a name without a new line to have the same representation. I propose to replace `\n` by a weird unicode symbol that is unlikely to happen, which reduces the interaction space here.

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/449/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/252/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@12c2e9b8e756548b4aeef743e758071acf82c3f7
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/449">pull/449/head       </A> | sbt/zinc@b27b222709f4ec408c84bd3154b5cc9bcd93fea6
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@e8c9757799bb0b251e5c5aefd96f841428d0eb9a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@91b56c50db93294003511fd40967c9296b07ef45
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@f4eadfc259f6f25b61eeb519fe29fb44220bec4a
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/449/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/253/console">is in progress...</a>

File: internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala (hunk 1 above)
Comment (eed3si9n) [reply internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala]: How about "♨0A" 
http://www.fileformat.info/info/unicode/char/2668/index.htm plus the hex code for `\n`.

Comment (jilen) [reply internal/zinc-core/src/main/scala/sbt/internal/inc/Relations.scala]: @jvican  @eed3si9n  Now `\n` is replaced with `\u2668` as @eed3si9n suggested

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/449/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/253/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@12c2e9b8e756548b4aeef743e758071acf82c3f7
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/449">pull/449/head       </A> | sbt/zinc@0fb6be6498e41696708d476c6efd6fd62b0dfb78
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@e8c9757799bb0b251e5c5aefd96f841428d0eb9a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@91b56c50db93294003511fd40967c9296b07ef45
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@f4eadfc259f6f25b61eeb519fe29fb44220bec4a
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/449/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

File: internal/zinc-core/src/main/scala/sbt/internal/inc/UsedName.scala [hunk 2]
```diff
@@ -17,6 +17,6 @@ object UsedName {
   def apply(name: String, scopes: Iterable[UseScope] = Nil): UsedName = {
     val useScopes = util.EnumSet.noneOf(classOf[UseScope])
     scopes.foreach(useScopes.add)
-    UsedName(name, useScopes)
+    UsedName(name.replaceAllLiterally("\n", "\u2668"), useScopes)
```
Comment (eed3si9n) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/UsedName.scala]: I think this should be "\u2668" + "0A" so if we come across another control code we can extend the solution.

**Comment by dwijnand** [comment]
How feasible is it to create a test representative of the downstream problem?

**Comment by jvican** [comment]
@jilen My feedback has not been fully addressed :wink:

**Comment by jilen** [comment]
@jvican 
 Where should I put the function ? Sorry, I am not quite familiar with sbt.
Is it ok to create an private function called `escapseControlChars` inside `UsedName` object ?

**Comment by jvican** [comment]
That's the ideal solution, thanks @jilen :smile:

**Comment by eed3si9n** [comment]
@dwijnand We have property-based testing in https://github.com/sbt/zinc/blob/1.x/internal/zinc-persist/src/test/scala/sbt/inc/AnalysisGenerators.scala.
Would adding `\n` in 

```scala
  def genScalaName: Gen[String] = Gen.listOf(Gen.choose('!', 'Z')).map(_.toString())
```

repro this issue?

**Comment by dwijnand** [comment]
I don't know but I like the sound of that.

**Comment by jilen** [comment]
@eed3si9n @jvican It is ready for review now. 
Actually this bug waste at least 30 minutes (due to recompiling the whole project) per day.
Hope it could be fixed soon both for sbt 1.0 and sbt 0.13.x

File: internal/zinc-core/src/main/scala/sbt/internal/inc/UsedName.scala [hunk 3]
```diff
@@ -14,9 +14,15 @@ import xsbti.UseScope
 case class UsedName(name: String, scopes: util.EnumSet[UseScope])
 
 object UsedName {
+
+  private def escapeControlChars(name: String) = {
+    name.replaceAllLiterally("\n", "\u26680A")
+  }
+
   def apply(name: String, scopes: Iterable[UseScope] = Nil): UsedName = {
+    val escapedName = escapeControlChars(name)
```
Comment (jvican) [inline internal/zinc-core/src/main/scala/sbt/internal/inc/UsedName.scala]: This is only done for the `apply` in `UsedName`. If there's some part of the code that uses the constructor of the case class directly, the name will not be escaped. Can you make sure that the constructor is not used and make it private?

File: internal/zinc-persist/src/test/scala/sbt/inc/AnalysisGenerators.scala [hunk 4]
```diff
@@ -170,7 +170,9 @@ trait AnalysisGenerators {
       external <- someOf(src.external.all.toList)
     } yield Relations.makeClassDependencies(Relation.empty ++ internal, Relation.empty ++ external)
 
-  def genScalaName: Gen[String] = Gen.listOf(Gen.choose('!', 'Z')).map(_.toString())
+  def genScalaName: Gen[String] = {
+    Gen.listOf(Gen.oneOf(Gen.choose('!', 'Z'), Gen.const('\n'))).map(_.toString())
```
Comment (jvican) [inline internal/zinc-persist/src/test/scala/sbt/inc/AnalysisGenerators.scala:174]: Did you try this before and after the change to make sure that it reproduced the bug?

**Review by jvican — CHANGES_REQUESTED** [review]
Hey @jilen, closer to the finish line!
  
Note: this fix will only be available in sbt 1.0. I think you should tell the Quill team that using newlines in names is a bad idea: it makes their code unfriendly to tools and unparseable.

File: internal/zinc-persist/src/test/scala/sbt/inc/AnalysisGenerators.scala (hunk 4 above)
Comment (jilen) [reply internal/zinc-persist/src/test/scala/sbt/inc/AnalysisGenerators.scala:174]: @jvican Yes

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/262/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/449/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/262/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@9232a8fbc872c2f94a91ec82d797a11140a271a0
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/449">pull/449/head       </A> | sbt/zinc@7f63a129f640a43b6e87716c74b9eb01894ee0f3
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@e8c9757799bb0b251e5c5aefd96f841428d0eb9a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@91b56c50db93294003511fd40967c9296b07ef45
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@f4eadfc259f6f25b61eeb519fe29fb44220bec4a
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:x: The result is: ***FAILED***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/449/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Comment by typesafe-tools** [comment]
A validation involving this pull request <a href="https://jenkins.scala-sbt.org:8592/job/sbt-validator/263/console">is in progress...</a>

**Comment by typesafe-tools** [comment]
[//]: # (Refs: 1.x-pull/449/head-1.x-1.x-1.x-1.x)
The validator <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/263/">has checked</A> the following projects against Scala 2.12,
tested using dbuild, projects built on top of each other.

Project             | Reference            | Commit
:------------------ | :------------------- | :----------------
  sbt               | <A HREF="https://github.com/sbt/sbt/tree/1.x">1.x                 </A> | sbt/sbt@9232a8fbc872c2f94a91ec82d797a11140a271a0
  zinc              | <A HREF="https://github.com/sbt/zinc/pull/449">pull/449/head       </A> | sbt/zinc@3a6c1149b045f78f5887bccce8116b02bb92e48b
  io                | <A HREF="https://github.com/sbt/io/tree/1.x">1.x                 </A> | sbt/io@e8c9757799bb0b251e5c5aefd96f841428d0eb9a
  librarymanagement | <A HREF="https://github.com/sbt/librarymanagement/tree/1.x">1.x                 </A> | sbt/librarymanagement@91b56c50db93294003511fd40967c9296b07ef45
  util              | <A HREF="https://github.com/sbt/util/tree/1.x">1.x                 </A> | sbt/util@f4eadfc259f6f25b61eeb519fe29fb44220bec4a
  website           | <A HREF="https://github.com/sbt/website/tree/1.x">1.x                 </A> | 

:white_check_mark: The result is: ***SUCCESS***
<A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild?sbt=1.x&zinc=pull/449/head&io=1.x&librarymanagement=1.x&util=1.x&website=1.x&deploy_to_bintray=false&run_sbt_community=false">(restart)</A>

**Review by dwijnand — APPROVED** [review]

**Review by jvican — APPROVED** [review]

**Comment by retronym** [comment]
This change represents a slight performance penalty: `replaceAllLiterally` calls `Matcher.quoteReplacement` each time on the name. 

`replace('\n', '\u26680A')` would be more efficient.

**Comment by jvican** [comment]
@jilen mind sending a PR to make that improvement?

**Comment by retronym** [comment]
https://github.com/sbt/zinc/pull/583

**Pull request merged**